x/ccv/provider/keeper/grpc_query.go

```diff
@@ -333,18 +335,63 @@ func (k Keeper) QueryConsumerValidators(goCtx context.Context, req *types.QueryC
 
 	ctx := sdk.UnwrapSDKContext(goCtx)
 
-	if _, found := k.GetConsumerClientId(ctx, consumerId); !found {
-		// chain has to have started; consumer client id is set for a chain during the chain's spawn time
-		return nil, status.Error(codes.InvalidArgument, fmt.Sprintf("no started consumer chain: %s", consumerId))
+	// get the consumer phase
+	phase, ok := k.GetConsumerPhase(ctx, consumerId)
+	if !ok {
+		return nil, status.Error(codes.InvalidArgument, errorsmod.Wrap(types.ErrUnknownConsumerId, consumerId).Error())
 	}
 
-	var validators []*types.QueryConsumerValidatorsValidator
+	// query consumer validator set
 
-	consumerValSet, err := k.GetConsumerValSet(ctx, consumerId)
-	if err != nil {
-		return nil, status.Error(codes.Internal, err.Error())
+	var consumerValSet []types.ConsensusValidator
+	var err error
+
+	// if the consumer launched, the consumer valset has been persisted
+	if phase == Launched {
+		consumerValSet, err = k.GetConsumerValSet(ctx, consumerId)
+		if err != nil {
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+
+		//  if the consumer hasn't launched yet, compute the consumer validator set
+	} else {
+		bondedValidators, err := k.GetLastBondedValidators(ctx)
+		if err != nil {
+			return nil, status.Error(codes.Internal, fmt.Sprintf("failed to get last validators: %s", err))
+		}
+		if topN := k.GetTopN(ctx, consumerId); topN > 0 {
+			// in a Top-N chain, we automatically opt in all validators that belong to the top N
+			// of the active validators
+			activeValidators, err := k.GetLastProviderConsensusActiveValidators(ctx)
+			if err != nil {
+				// something must be broken in the bonded validators, so we have to panic since there is no realistic way to proceed
+				return nil, status.Error(codes.Internal, fmt.Sprintf("failed to get active validators: %s", err))
+			}
+
+			minPower, err := k.ComputeMinPowerInTopN(ctx, activeValidators, topN)
+			if err != nil {
+				// we panic, since the only way to proceed would be to opt in all validators, which is not the intended behavior
+				return nil, status.Error(codes.Internal, fmt.Sprintf("failed to compute min power to opt in for chain %s: %s", consumerId, err))
+			}
+
+			// set the minimal power of validators in the top N in the store
+			k.SetMinimumPowerInTopN(ctx, consumerId, minPower)
+
+			k.OptInTopNValidators(ctx, consumerId, activeValidators, minPower)
+		}
+
+		consumerValSet = k.ComputeNextValidators(ctx, consumerId, bondedValidators)
+
+		// sort the address of the validators by ascending lexical order as they were persisted to the store
+		sort.Slice(consumerValSet, func(i, j int) bool {
```

Comment:
Why do we need the sort?

Reply:
To keep the returned validator order consistent regardless of whether the consumer has launched or not.
It also simplifies the testing.